core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -360,6 +371,17 @@ class BlockManagerMasterEndpoint(
 
   }
 
+  private def addMergerLocation(blockManagerId: BlockManagerId): Unit = {
+    if (!shuffleMergerLocations.contains(blockManagerId.host) && !blockManagerId.isDriver) {
+      val shuffleServerId = BlockManagerId(blockManagerId.executorId, blockManagerId.host,
+        StorageUtils.externalShuffleServicePort(conf))
+      if (shuffleMergerLocations.size >= maxRetainedMergerLocations) {
+        shuffleMergerLocations -= shuffleMergerLocations.head._1
```

Comment:
hmm..why we need to set a threshold of the `shuffleMergerLocations`? I think the number of external shuffle service should be static within a cluster(no matter what the resource allocation mode of the application is, dynamic or static) and wouldn't increase unlimitedly.

Besides, the removed oldest merger may store more merged shuffle data than others...would be better if we could remove depends on merged shuffle data size...but this should be another topic though..

Reply:
This would be useful in a cloud based deployment where the nodes keeps coming up and going down and so the number of External shuffle services for a cluster is not static. 

> Besides, the removed oldest merger may store more merged shuffle data than others...would be better if we could remove depends on merged shuffle data size...but this should be another topic though..
Agree, the first part of that is to have a pluggable API (https://issues.apache.org/jira/browse/SPARK-33329) that way we can come up with specific implementation based on cloud/on-prem/ as well as based on scheduler backends. This is the basic implementation but eventually we need load balancing between shuffle mergers

Reply:
Just to clarify, removing a merger does not remove the merged shuffle data on that location. It only prevents using that location for future shuffles. Reducers will still be able to fetch merged shuffle data from the removed merger locations.